src/julia_gc.cc

```diff
@@ -1089,3 +1381,61 @@ void MarkJuliaWeakRef(void * p)
     if (JMarkTyped(p, jl_weakref_type))
         YoungRef++;
 }
+
+
+#ifdef TEST_JULIA_GC_INTERNALS
+
+struct IntCmp {
+    static int cmp(int a, int b)
+    {
+        return a - b;
+    }
+};
+
+__attribute__((constructor)) void TestScapegoatTree()
+{
+    const Int                    N = 1024 * 1024;
+    BalancedTree<int, IntCmp> * btree = BalancedTree<int, IntCmp>::make();
+    btree->init();
+    for (int i = 0; i < N; i++) {
+        btree->insert(i);
+    }
+    if (btree->count() != N)
+        abort();
+    if (height_to_size[btree->depth() - 1] > btree->count())
+        abort();
+    for (int i = 0; i < N; i++) {
+        if (!btree->find(i))
+            abort();
+    }
+    for (int i = 1; i < N; i++) {
+        btree->remove(i);
+    }
+    if (btree->count() != 1 || btree->depth() != 1)
+        abort();
+    btree->remove(0);
+    for (int i = 0, j = 0; i < N; i++, j = (5 * j + 1) & (N - 1)) {
+        btree->insert(j);
+    }
+    if (btree->count() != N)
+        abort();
+    if (height_to_size[btree->depth() - 1] > btree->count())
+        abort();
+    for (int i = 0; i < N; i++) {
+        if (!btree->find(i))
+            abort();
+    }
+    for (int i = 0, j = 0; i < N; i++, j = (j + 7) & (N - 1)) {
+        if (i == N / 2 && btree->count() != N / 2)
+            abort();
+        btree->remove(j);
+    }
+    if (btree->count() != 0 || btree->depth() != 0)
+        abort();
+    btree->remove(0);
+    BalancedTree<int, IntCmp>::destroy(btree);
+    printf("Scapegoat tree passed tests.\n");
+    exit(0);
```

Comment:
There are no tests for the `ArrayList`, it seems?

Reply:
No. I copied these tests over from my original implementation, which did not have any for `ArrayList`, which is a pretty barebones dynamic array.